### ISSUE TITLE
Add units/coordinate system support?

### ISSUE THREAD

**Issue by ianmackenzie** [ianmackenzie/elm-geometry#66]
Currently, `elm-geometry` has no explicit support for tracking what units or coordinate systems any given value is defined in; it's up to you to remember that in (for example)

```elm
Point2d.fromCoordinates ( 100, 200 )
```

the coordinate values refer to numbers of pixels, and the point is defined in a coordinate system where (0, 0) is at the bottom-left corner of some SVG drawing and positive Y points up. This allows for several kinds of subtle errors, especially when conversions between units or coordinate systems are involved; for example, if you're doing some geometric modelling in real-world coordinates (meters, feet etc.) and then converting to dimensions in pixels for rendering, or doing most drawing work in a Y-up coordinate system before converting to a Y-down coordinate system for final rendering.

In Elm 0.18 it would have been prohibitively expensive to track units of individual `Float` values since this would have required wrapping them in some sort of opaque type with an extra type variable, and that would mean a dynamic memory allocation for every single `Float` value! However, in Elm 0.19, union types with a single constructor that has a single argument have a runtime representation of just that argument, so something like

```elm
type WithUnits units =
    WithUnits Float
```

will get compiled down to just a plain `Float` at runtime. (Note that the `units` type variable isn't actually used as part of the representation - it's used only for type checking at compile time.) As a result, I've been experimenting with an [`elm-units`](https://github.com/ianmackenzie/elm-units) package that does exactly that, to allow you to write things like `meters 3.6` or `pixels 100` (or even `milliseconds 500` or `milesPerHour 60`) and have them properly type-checked, with any necessary conversions happening automatically (but with little to no runtime overhead).

Once `elm-units` is published I would very much like to switch `elm-geometry` over to storing things like point coordinates as `Length` values instead of raw `Float`s. At the same time, I'd also like to add support for tracking which coordinate system any given value is defined in. I think in this world, defining points would look like

```elm
module Point2d exposing (..)

fromCoordinates -> ( Length space, Length space ) -> Point2d space coordinateSystem
```
A `space` basically defines a system of units - `elm-units` defines the `InWorld` space type where units are meters, feet, inches etc., and `OnScreen` where units are pixels, but you can define your own like `InGameWorld` where perhaps things are measured in numbers of game tiles. The `coordinateSystem` then specifies what coordinate system those coordinates are measured in. (Top left of the screen, with Y pointing down? Bottom left, with Y pointing up?)

Note that the `space` of the resulting point is determined by the types of coordinates you pass in, but `coordinateSystem` is left as a free type variable that can take on any value you want. You can then declare that a point is defined in a particular coordinate system simply by adding a type annotation! This means you can do a bunch of work inside a function (creating a bunch of points, building up geometry from them) and then add a type annotation just to the final result that declares what coordinate system it is defined in.

You'd use `Point2d.fromCoordinates` like this:

```elm
module MyApp

import Length exposing (pixels)
import Point2d exposing (Point2d)

-- Define a coordinate system by making a unique type 'tag' for it
type YUpCoordinates =
    YUpCoordinates

-- Create a point by simply declaring it to be in a particular coordinate system
-- (OnScreen is inferred since 'pixels 200' has type 'Length OnScreen')
svgPoint : Point2d OnScreen YUpCoordinates
svgPoint =
    Point2d.fromCoordinates ( pixels 200, pixels 100 )
```

`Frame2d` and `Frame3d` values would then have type variables indicating both what coordinate system they are _defined in_ as well as what coordinate system (if any) they themselves _define_:

```elm
module Frame2d

type DefinedIn globalCoordinates
    = DefinedIn Never
    
type Defines localCoordinates
    = Defines Never

atPoint :
    Point2d space globalCoordinates
    -> Frame2d space (DefinedIn globalCoordinates) (Defines localCoordinates)
```

Similar to points, what coordinate system a frame is defined in will often be determined by the points/directions you construct it from, but you can declare that it defines a particular coordinate system by adding a type annotation:

```elm
module MyApp

drawingHeight : Length OnScreen
drawingHeight =
    pixels 400

-- Top-left corner of an SVG drawing, defined in Y-up coordinates relative to the
-- bottom left corner
topLeftPoint : Point2d OnScreen YUpCoordinates
topLeftPoint =
    Point2d.fromCoordinates ( pixels 0, drawingHeight )

-- 'OnScreen' and 'DefinedIn YUpCoordinates' are inferred from 'topLeftPoint'; a type
-- annotation is used to declare that this frame defines the Y-down coordinate system
topLeftFrame : Frame2d OnScreen (DefinedIn YUpCoordinates) (Defines YDownCoordinates)
topLeftFrame =
    Frame2d.atPoint topLeftPoint |> Frame2d.reverseY
```

The `placeIn` and `relativeTo` functions would be upgraded to track coordinate systems properly:

```elm
module Point2d exposing (..)

relativeTo :
    Frame2d space (DefinedIn globalCoordinates) (Defines localCoordinates)
    -> Point2d space (DefinedIn globalCoordinates)
    -> Point2d space (DefinedIn localCoordinates)

placeIn :
    Frame2d space (DefinedIn globalCoordinates) (Defines localCoordinates)
    -> Point2d space (DefinedIn localCoordinates)
    -> Point2d space (DefinedIn globalCoordinates)
```

This means you could draw something in Y-up coordinates, then finally convert to Y-down, with the compiler checking that everything makes sense:

```elm
triangleInYUp : Triangle2d OnScreen YUpCoordinates
triangleInYUp =
    let
        p1 =
            Point2d.fromCoordinates ( pixels 100, pixels 100 )
            
        p2 =
            Point2d.fromCoordinates ( pixels 300, pixels 100 )
            
        p3 =
            Point2d.fromCoordinates ( pixels 100, pixels 400 )
    in
    Triangle2d.fromVertices ( p1, p2, p3 )

triangleInYDown : Triangle2d OnScreen YDownCoordinates
triangleInYDown =
    triangleInYUp |> Triangle2d.relativeTo topLeftFrame
```

In general, type annotations will get more verbose (but more descriptive!), but actual function bodies/implementations won't change much:

```elm
p1 : Point3d InWorld SiteCoordinates
p1 =
    Point3d.fromCoordinates ( meters 20, meters 30, meters 0 )
    
p2 : Point3d InWorld SiteCoordinates
p2 =
    -- You can mix and match units as long as they're in the same space!
    Point3d.fromCoordinates ( centimeters 12, feet 36.2, inches 5 )
    
displacement : Vector2d (LengthUnits InWorld) SiteCoordinates
displacement =
    Vector2d.from p1 p2
```

Note that although I typed it out in the type annotation, the type of `displacement` including its units and coordinate system can be fully inferred from the types of `Vector2d.from`, `p1` and `p2`; `Vector2d.from` would have the type signature

```elm
module Vector2d

from :
    Point2d space coordinateSystem
    -> Point2d space coordinateSystem
    -> Vector2d (LengthUnits space) coordinateSystem 
```

Whew! Thanks for reading this far =) I'm really excited about the improved compile-time error checking these changes would bring for the types of applications I'm personally most interested in (designing complex, safety-critical 3D things like bridges, buildings, cars, robots, rockets or space stations), but I realize those goals can conflict with `elm-geometry` being a good lightweight general-purpose geometry library for other things like on-screen diagrams/animations or games. So I really wanted to get people's thoughts before I dove into this! Specifically I'd love to get your feedback on:

  - What's your reaction to the idea in general?
    - "That would be fantastic"?
    - "Sure, makes sense"?
    - "Whoa, `elm-geometry` is too heavyweight/verbose already"?
  - Are there examples in your own code where it's not clear how it could be updated to this design, for example cases where it's not obvious what the units or coordinate system would be?
  - Any thoughts/questions on the related `elm-units` package?

**Comment by acateland** [comment]
in my actual use case, "sure make sense", i guess for many people it will be "That would be fantastic"

**Comment by ianmackenzie** [comment]
Thanks @acateland! And yes, the Mars orbiter is definitely a motivating example =) Glad the additional verbosity didn't seem *too* onerous...

**Comment by mpizenberg** [comment]
I had a look at elm-units when I saw it in my home page thread ^^. I think it's a very good idea in general. I'm a bit more reserved about the implications it would have on using a package like elm-geometry. That's just my gut feeling though, will have to think about it more.

**Comment by ianmackenzie** [comment]
@mpizenberg sounds good! I'm fairly nervous about the (potential) change myself - there are some definite downsides - so I'm not at all surprised you have some reservations. I look forward to hearing your thoughts once you've had some time to mull it over! And I'm happy to sketch out what I think particular functions/use cases might look like with units/coordinate systems, if you get to a point where you're wondering "but what would *this* kind of code end up looking like"...

**Comment by xarvh** [comment]
My first reaction was "Sure, makes sense".

I don't have code with elm-geometry; thinking about using it or `elm-units` inside my game I wonder if I will have to replace a lot of operators with functions... Adding two angles goes from `a + b` to `Angle.add a b`.
It's not the end of the world, but I'm not sure it would make the code more readable.

If I understand correctly, the only practical changes for the user are 1) more verbose types and 2) can't use ops for 1-dimensional quantities, neither is too bad and might be worth the safety.

The main problem I see is that we don't have enough use cases, no major project that uses the library, so it's difficult to decide which path to take.
On the other hand, this means that if it turns out to be a bad idea, you can afford to remove the feature from the next version without damaging too much anyone.

I think you should try.
While the benefits for elm-geometry are not clear cut, this is uncharted territory for Elm and you'll definitely discover a lot of stuff that will benefit the language and the community.

TL;DR: we don't know for certain that it will be good, but we know for certain that we will learn *a lot* from this, and this makes it worth trying.

**Comment by mpizenberg** [comment]
One of the things I'm worried about, is that in code that include lots of maths, we do many mistakes that are not directly linked to types or units. Often, having few Debug.log around, can help figure out why algorithms are not working. Now if we rely on the fact that the --optimize performance is ok, but the normal one is terrible, that might make debugging much less pleasant. All this is very hypothetical but maybe it's worth exploring in a separate package? A bit like stylish-elephants that is the chosen one now. This would be a lot of work of course and I know Ian you're already spending a lot of time in elm-geometry so I definitely don't want to double the weight.

**Comment by ianmackenzie** [comment]
Debug mode performance is a very good point @mpizenberg! I think there are might be some mitigating techniques though - for example it's possible that a `Point3d` could remain a plain `( Float, Float, Float )` _internally_, converting back and forth to `( Length space, Length space, Length space )` only when `coordinates` or `fromCoordinates` are called. That way all the internal code could just be based on plain `Float` values. (That's actually a tempting option for other reasons!)

If I start down this road I think I'll do it in a branch anyways, which should make it fairly straightforward to publish as a separate experimental package if desired.

**Comment by ianmackenzie** [comment]
So one possibility (spoiler alert: I don't really like it) is that `elm-geometry` could be split into one full-blown heavyweight package with opaque types, units and coordinate systems, and a lighter-weight package that `elm-geometry` itself would probably depend on. My impression is that in many of the cases where a lighter-weight package would be desirable it's because you want to use some of the geometric _functions_ provided by `elm-geometry` (polygon triangulation, curve evaluation, bounding box operations etc) but you don't want to commit to using the `elm-geometry` _types_ everywhere.

Right now it's a bit verbose to convert back and forth using `Point2d.fromCoordinates`, `Point2d.coordinates` etc., so that is unappealing. What if there were separate `elm-2d-math` and `elm-3d-math` packages that provided much of the actual functionality of `elm-geometry` but just using plain `Float`s, tuples and records? `elm-geometry` could then build on top of those by using those very same tuples and records wrapped inside opaque types that enforced the point/vector/direction distinction, tracked units and coordinate systems, etc. But if you just wanted to work on `( Float, Float )` values directly then you could use `elm-2d-math` (knowing that the semantics would be pretty much the same as `elm-geometry`, so it would be fairly straightforward to migrate from one to the other if needed).

I've sketched up the beginnings of a possible `elm-2d-math` API [here](https://gist.github.com/ianmackenzie/14e1d1f275b476cd05a68aed2e08af46) - thoughts? I personally really don't like the idea of having two separate packages that basically do the same thing - I'm like 75% sure this is the wrong approach - but I'm interested in hearing everyone's thoughts!

**Comment by architectcodes** [comment]
For the record: I had a long chat with @ianmackenzie about integrating elm-units with elm-geometry yesterday and I’m incredibly excited about it. It’s not even "That would be fantastic" on my end. It’s more like ”FINALLY!!!! I feel like I’ve been waiting for this all my life!!!!“

Thing is, in a CAD-like context units are incredibly important. In a primitive environment like AutoCAD (originally released in the early 80s and hardly changed since then), everything was based on unit-less values. It was up to the user what convention they pick, which units represent what measurements etc. And it‘s always been a big usability problem! Scaling issues are pretty common when working on top of AutoCAD file formats.

Newer CAD formats have units baked in and it’s a huge blessing. Everything is explicit. There are no compatibility issues. No user-specific conventions. No problems with things like importing geometry developed by someone else. (Think: third-party libraries)

Back to the context of Elm programming, in the project of mine that uses elm-geometry I actually have an entire collection of Elm modules that deal specifically with conversions between units. Every time when using elm-geometry, I need to remember to convert the units I use in my model to millimeters. It’s a nuisance and pretty error prone!

Millimeters because that’s the convention I picked for my usage of elm-geometry. If I were to integrate a library developed by someone else (with, say, furniture), I bet I’d have compatibility issues because the other person decided to use inches rather than millimeters as the base unit. Or feet. Or meters. Or centimeters. You see what I mean?

So I’m incredibly excited about integrating elm-units into elm-geometry. Even if that means prioritising API usability over performance (which I don’t even think it does!) I’d be all in for it. It prevents so many potential mistakes!

**Comment by mpizenberg** [comment]
I totally get that, for anything related to physics, industrial design, CAD, ... units are super important. I think it's just that we had different use cases in mind. To me geometry is a math concept that do not need a unit system. Most important things are coordinate systems but having a unit system that correctly encapsulates the transformability of coordinate systems without having a too big written and mental overhead is probably a challenge.
But in the meantime, I've never encountered yet a geometry/algebra library with units baked in, so I'm curious to what this could bring. We will never know if we don't try so why not try :)

**Comment by ianmackenzie** [comment]
Nice article by @joelq on (among other things) the value of separating 'game coordinates' from 'screen coordinates': https://robots.thoughtbot.com/decoupling-all-the-things

**Comment by ianmackenzie** [comment]
Update is now underway in the [`coordinate-systems`](/ianmackenzie/elm-geometry/tree/coordinate-systems) branch, tracked in #86.
